Mathlib/CategoryTheory/EffectiveEpi/Enough.lean

```diff
@@ -0,0 +1,77 @@
+/-
+Copyright (c) 2024 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson
+-/
+import Mathlib.CategoryTheory.EffectiveEpi.Basic
+/-!
+
+# Effectively enough objects in the image of a functor
+
+We define the class `F.EffectivelyEnough` on a functor `F : C ⥤ D` which says that for every object
+in `D`, there exists an effective epi to it from an object in the image of `F`.
+
+## TODO
+* Show that `Stonean.toCompHaus` and `Stonean.toProfinite` satisfy this property.
+* Use this as one of the sufficient conditions for a functor to induce an equivalence of sheaf
+  categories of coherent and regular sheaves.
+-/
+
+namespace CategoryTheory
+
+open Limits
+
+variable {C D : Type*} [Category C] [Category D] (F : C ⥤ D)
+
+namespace Functor
+
+/--
+An effective presentation of an object `X` with respect to a functor `F` is the data of an effective
+epimorphism of the form `F.obj p ⟶ X`.
+-/
+structure EffectivePresentation (X : D) where
+  /-- The object of `C` giving the source of the effective epi -/
+  p : C
+  /-- The morphism `F.obj p ⟶ X` -/
+  f : F.obj p ⟶ X
+  /-- `f` is an effective epi -/
+  effectiveEpi : EffectiveEpi f
+
+/--
+`D` has *effectively enough objects with respect to the functor `F` if every object has an
+effective presentation.
+-/
+class EffectivelyEnough : Prop where
+  /-- For every `X : D`, there exists an object `p` of `C` with an effective epi `F.obj p ⟶ X`. -/
+  presentation : ∀ (X : D), Nonempty (F.EffectivePresentation X)
+
+variable [F.EffectivelyEnough]
+
+/--
+`F.objOver X` provides an arbitrarily chosen object in the image of `F` equipped with an effective
+epimorphism `F.morphismOver : F.objOver X ⟶ X`.
+-/
+noncomputable def objOver (X : D) : D :=
+  F.obj (EffectivelyEnough.presentation (F := F) X).some.p
+
+/--
+The epimorphism `F.morphismOver : F.objOver X ⟶ X` from the arbitrarily chosen object in the image
+of `F` over `X`.
+-/
+noncomputable def morphismOver (X : D) : F.objOver X ⟶ X :=
```

Comment:
I think the following name would be more specific:
```suggestion
noncomputable def effectiveEpiOver (X : D) : F.objOver X ⟶ X :=
```
Then, `objOver` could be renamed `effectiveEpiOverObj`.